### PR TITLE
Added validation for store id in CMS Block

### PR DESCRIPTION
**Pull request by Rus0**
### Description (*)

Adding a filter for store id in the blocks, using the store send in the headers

### Fixed Issues (if relevant)

1. magento/graphql-ce#691: cmsBlocks that has scope limited to specific Store View can be seen on other Store Views in case there are Blocks with identical identifiers on both Store Views

### Manual testing scenarios (*)

1. All the steps in the ticket #691 

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds are green)

### CODE REVIEW

File: app/code/Magento/CmsGraphQl/Model/Resolver/DataProvider/Block.php [hunk 1]
```diff
@@ -27,29 +31,67 @@ class Block
...
      * @param string $blockIdentifier
      * @return array
-     * @throws NoSuchEntityException
+     * @throws LocalizedException
      */
     public function getData(string $blockIdentifier): array
     {
-        $block = $this->blockRepository->getById($blockIdentifier);
+        $filterBy = BlockInterface::IDENTIFIER;
+        $storeId = (int)$this->storeManager->getStore()->getId();
```
Comment (lenaorobei) [inline app/code/Magento/CmsGraphQl/Model/Resolver/DataProvider/Block.php]: Store needs to be passed to this method from resolver `$context->getExtensionAttributes()->getStore()`.

File: app/code/Magento/CmsGraphQl/Model/Resolver/DataProvider/Block.php [hunk 2]
```diff
@@ -27,29 +31,67 @@ class Block
...
+        if (is_numeric($blockIdentifier)) {
+            $filterBy = BlockInterface::BLOCK_ID;
+        }
+        $searchCriteria = $this->searchCriteriaBuilder->addFilter(
```
Comment (lenaorobei) [inline app/code/Magento/CmsGraphQl/Model/Resolver/DataProvider/Block.php]: Please use collection and `getFirstItem` here.

**Review by lenaorobei — CHANGES_REQUESTED** [review]
Please see my comments.

File: app/code/Magento/CmsGraphQl/Model/Resolver/DataProvider/Block.php [hunk 3]
```diff
@@ -18,38 +25,56 @@
...
+        $storeId = (int)$currentStore->getId();
+        if (is_numeric($blockIdentifier)) {
+            $filterBy = BlockInterface::BLOCK_ID;
+        }
+
+        /** @var BlockCollection $collection */
+        $collection = $this->blockCollectionFactory->create();
+        $collection->addFieldToFilter($filterBy, ["eq" => $blockIdentifier]);
+        $collection->addFieldToFilter("store_id", ["eq" => $storeId]);
+        $collection->load();
```
Comment (lenaorobei) [inline app/code/Magento/CmsGraphQl/Model/Resolver/DataProvider/Block.php]: Please use `->getFirstItem()` and check `->getSize() > 0`.

Comment (lenaorobei) [inline app/code/Magento/CmsGraphQl/Model/Resolver/DataProvider/Block.php]: You can safely remove
```
$collection->load();
        if ($collection->count() < 1) {
            throw new NoSuchEntityException(
                __('The CMS block with the "%1" ID doesn\'t exist.', $blockIdentifier)
            );
        }
```
`getFirstItem` will load the collection if needed.

File: app/code/Magento/CmsGraphQl/Model/Resolver/Blocks.php [hunk 4]
```diff
@@ -74,16 +77,16 @@ private function getBlockIdentifiers(array $args): array
...
-    private function getBlocksData(array $blockIdentifiers): array
+    private function getBlocksData(array $blockIdentifiers, Store $currentStore): array
     {
         $blocksData = [];
         foreach ($blockIdentifiers as $blockIdentifier) {
             try {
-                $blocksData[$blockIdentifier] = $this->blockDataProvider->getData($blockIdentifier);
-            } catch (NoSuchEntityException $e) {
+                $blocksData[$blockIdentifier] = $this->blockDataProvider->getData($blockIdentifier, $currentStore);
+            } catch (LocalizedException $e) {
```
Comment (lenaorobei) [inline app/code/Magento/CmsGraphQl/Model/Resolver/Blocks.php]: Why did you change to `LocalizedException`?

File: app/code/Magento/CmsGraphQl/Model/Resolver/DataProvider/Block.php [hunk 5]
```diff
@@ -9,7 +9,14 @@
 
 use Magento\Cms\Api\BlockRepositoryInterface;
```
Comment (lenaorobei) [inline app/code/Magento/CmsGraphQl/Model/Resolver/DataProvider/Block.php]: Please remove redundant dependencies.

File: app/code/Magento/CmsGraphQl/composer.json [hunk 6]
```diff
@@ -5,6 +5,7 @@
     "require": {
         "php": "~7.1.3||~7.2.0||~7.3.0",
         "magento/framework": "*",
+        "magento/module-store": "*",
```
Comment (lenaorobei) [inline app/code/Magento/CmsGraphQl/composer.json:8]:
```suggestion
        "magento/module-store": "*",
```

**Review by lenaorobei — CHANGES_REQUESTED** [review]
Please see address comments.

File: app/code/Magento/CmsGraphQl/Model/Resolver/Blocks.php (hunk 4 above)
Comment (Rus0) [reply app/code/Magento/CmsGraphQl/Model/Resolver/Blocks.php]: Good question, I do not remember hahaha, changing it back

File: app/code/Magento/CmsGraphQl/composer.json (hunk 6 above)
Comment (Rus0) [reply app/code/Magento/CmsGraphQl/composer.json:8]: ??? I don't see any changes or comment @lenaorobei

Comment (lenaorobei) [reply app/code/Magento/CmsGraphQl/composer.json:8]: `magento/module-store` is redundant dependency

Comment (Rus0) [reply app/code/Magento/CmsGraphQl/composer.json:8]: If I don't set it, I get the error: Module Magento\CmsGraphQl has undeclared dependencies: hard [Magento\Store]

**Comment by lenaorobei** [comment]
@magento run Functional Tests build

**Review by lenaorobei — APPROVED** [review]

**Comment by magento-engcom-team** [comment]
Hi @lenaorobei, thank you for the review.
[ENGCOM-5826](https://jira.corp.magento.com/browse/ENGCOM-5826) has been created to process this Pull Request
:eight_spoked_asterisk: @lenaorobei, could you please add one of the following labels to the Pull Request?

| Label | Description |
|--------------------------------|-----------------------------------------------------------------|
| ```Auto-Tests: Covered``` | All changes in Pull Request is covered by auto-tests |
| ```Auto-Tests: Not Covered``` | Changes in Pull Request requires coverage by auto-tests |
| ```Auto-Tests: Not Required``` | Changes in Pull Request does not require coverage by auto-tests |

**Review by lenaorobei — APPROVED** [review]

**Comment by magento-engcom-team** [comment]
Hi @lenaorobei, thank you for the review.
[ENGCOM-5826](https://jira.corp.magento.com/browse/ENGCOM-5826) has been created to process this Pull Request

**Review by lenaorobei — APPROVED** [review]

**Comment by magento-engcom-team** [comment]
Hi @lenaorobei, thank you for the review.
[ENGCOM-5826](https://jira.corp.magento.com/browse/ENGCOM-5826) has been created to process this Pull Request

**Comment by TomashKhamlai** [comment]
Could you please merge the latest mainline to this PR?

In production mode, I am getting this error on schema loading
```
"debugMessage":"User Deprecated Functionality: GraphQL\\Type\\Definition\\Type::getInternalTypes is deprecated. Use Type::getStandardTypes()
```
I have seen this error several times on branches that are oudated.

**Comment by magento-cicd2** [comment]
[CLA assistant check](https://cla.dev.magento.com/magento/graphql-ce?pullRequest=870) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla.dev.magento.com/magento/graphql-ce?pullRequest=870) before we can accept your contribution.<br/>**2** out of **3** committers have signed the CLA.<br/><br/>:white_check_mark: Rus0<br/>:white_check_mark: lenaorobei<br/>:x: devilbox<hr/>**devilbox** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).

**Comment by TomashKhamlai** [comment]
Well done! :+1:

### Checked. Works as expected
:heavy_check_mark: The **title** is according to Store headers
:heavy_check_mark: The **content** is according to Store headers
:heavy_check_mark: Variables which got values in the Store View scope are displayed as expected
:heavy_check_mark: Variables which were set in the global scope are displayed as expected

### The problem
:x: If the block has the scope **All store views** it is not found.

### Details of the problem
#### Request 1
```graphql
query showCmsBlock(
  $identifiers: [String]
) {
  storeConfig {
    code
  }
  cmsBlocks(
    identifiers: $identifiers
  ) {
    items {
      content
      title
      identifier
    }
  }
}
```
#### Headers 1
No headers

#### Variables 1
```json
{
  "identifiers": ["test-block-2"]
}
```

#### Response 1
```graphql
{
  "errors": [
    {
      "message": "The CMS block with the \"test-block-2\" ID doesn't exist.",
      "extensions": {
        "category": "graphql-no-such-entity"
      },
      "locations": [
        {
          "line": 6,
          "column": 5
        }
      ],
      "path": [
        "cmsBlocks",
        "items",
        0
      ]
    }
  ],
  "data": {
    "storeConfig": {
      "code": "default"
    },
    "cmsBlocks": {
      "items": [
        null
      ]
    }
  }
}
```

#### Request 2
```graphql
query showCmsBlock(
  $identifiers: [String]
) {
  storeConfig {
    code
  }
  cmsBlocks(
    identifiers: $identifiers
  ) {
    items {
      content
      title
      identifier
    }
  }
}
```
#### Headers 2
```json
{
  "Store": "germstv"
}
```

#### Variables 2
```json
{
  "identifiers": ["test-block-2"]
}
```

#### Response 2
```graphql
{
  "errors": [
    {
      "message": "The CMS block with the \"test-block-2\" ID doesn't exist.",
      "extensions": {
        "category": "graphql-no-such-entity"
      },
      "locations": [
        {
          "line": 6,
          "column": 5
        }
      ],
      "path": [
        "cmsBlocks",
        "items",
        0
      ]
    }
  ],
  "data": {
    "storeConfig": {
      "code": "germstv"
    },
    "cmsBlocks": {
      "items": [
        null
      ]
    }
  }
}
```
#### Blocks enabled
image

#### Block that was requested
image

**Review by lenaorobei — CHANGES_REQUESTED** [review]
@Rus0, could you please address found issues?

**Comment by Rus0** [comment]
Sure, I will later this week

**Comment by lenaorobei** [comment]
@Rus0 unfortunately, I need to close this PR since there is no activity. Please feel free to reopen if you want to continue working on it. Thank you.
